File: swift/Tests/WarehouseTests/PlaceOrderTests.swift
import Testing
import Warehouse

@Suite("PlaceOrder")
struct PlaceOrderTests {
  @Test func happyPathPersistsPaid() throws {
    let (useCase, payments, repository, inventory) = try pipeline(stock: ["SKU-1": 10])
    let result = useCase.execute(
      lines: [try line("SKU-1", quantity: 2, minorUnits: 500)],
      idempotencyKey: "idem-1"
    )
    guard case .success(let order) = result else {
      Issue.record("expected success, got \(result)")
      return
    }
    #expect(order.status == .paid)
    #expect(order.id.value == "ord-1")
    let total = try order.total()
    let expected = try usd(1000)
    #expect(total == expected)
    #expect(payments.chargedOrders.count == 1)
    let stored = repository.get(order.id)
    #expect(stored?.status == .paid)
    #expect(inventory.snapshotStock()[try Sku("SKU-1")] == 8)
  }

  @Test func insufficientStockFailsWithoutChargeOrPersist() throws {
    let (useCase, payments, repository, inventory) = try pipeline(stock: ["SKU-1": 1])
    let result = useCase.execute(
      lines: [try line("SKU-1", quantity: 5, minorUnits: 500)],
      idempotencyKey: "idem-2"
    )
    guard case .failure(.insufficientStock(let error)) = result else {
      Issue.record("expected insufficient stock, got \(result)")
      return
    }
    #expect(error.available == 1)
    #expect(payments.chargedOrders.isEmpty)
    #expect(repository.saved.isEmpty)
    #expect(inventory.snapshotStock()[try Sku("SKU-1")] == 1)
  }

  @Test func paymentDeclineReleasesReservation() throws {
    let (useCase, payments, repository, inventory) = try pipeline(
      stock: ["SKU-1": 10],
      decline: true
    )
    let result = useCase.execute(
      lines: [try line("SKU-1", quantity: 1, minorUnits: 500)],
      idempotencyKey: "idem-3"
    )
    guard case .failure(.paymentDeclined(_)) = result else {
      Issue.record("expected payment declined, got \(result)")
      return
    }
    #expect(payments.chargedOrders.count == 1)
    #expect(repository.saved.isEmpty)
    #expect(inventory.snapshotStock()[try Sku("SKU-1")] == 10)
  }

  @Test func saveFailureRefundsAndReleases() throws {
    let (useCase, payments, repository, inventory) = try pipeline(stock: ["SKU-1": 10])
    repository.failSave = true
    let result = useCase.execute(
      lines: [try line("SKU-1", quantity: 1, minorUnits: 500)],
      idempotencyKey: "idem-4"
    )
    guard case .failure(.persistenceConflict(_)) = result else {
      Issue.record("expected persistence conflict, got \(result)")
      return
    }
    #expect(payments.refunded.count == 1)
    #expect(inventory.snapshotStock()[try Sku("SKU-1")] == 10)
  }

  @Test func compensationFailureAfterSaveFailure() throws {
    let (useCase, payments, repository, _) = try pipeline(stock: ["SKU-1": 10])
    repository.failSave = true
    payments.failRefund = true
    let result = useCase.execute(
      lines: [try line("SKU-1", quantity: 1, minorUnits: 500)],
      idempotencyKey: "idem-5"
    )
    guard case .failure(.compensationFailure(_)) = result else {
      Issue.record("expected compensation failure, got \(result)")
      return
    }
  }

  @Test func invalidLinesFailValidation() throws {
    let (useCase, _, _, _) = try pipeline(stock: [:])
    let result = useCase.execute(lines: [], idempotencyKey: "idem-6")
    guard case .failure(.invalidOrder(_)) = result else {
      Issue.record("expected invalid order, got \(result)")
      return
    }
  }

  @Test func getReturnsNilForUnknownId() throws {
    let (_, _, repository, _) = try pipeline(stock: ["SKU-1": 10])
    #expect(repository.get(try OrderId("missing-order")) == nil)
  }

  @Test func idempotentReplayDoesNotDoubleCharge() throws {
    var mapped: [Sku: Int] = [:]
    mapped[try Sku("SKU-1")] = 10
    let inventory = InMemoryInventoryGateway(stock: mapped)
    let payments = FakePaymentProcessor()
    let repository = InMemoryOrderRepository()
    let useCase = PlaceOrderUseCase(
      inventory: inventory,
      payments: payments,
      repository: repository,
      ids: SequenceOrderIdGenerator()
    )
    let lines = [try line("SKU-1", quantity: 2, minorUnits: 300)]
    let first = useCase.execute(lines: lines, idempotencyKey: "idem-7")
    let second = useCase.execute(lines: lines, idempotencyKey: "idem-7")
    guard case .success(_) = first, case .success(_) = second else {
      Issue.record("expected two successes, got \(first) \(second)")
      return
    }
    #expect(payments.chargedOrders.count == 1)
  }

  @Test func reusingKeyWithDifferentPayloadIsInvalid() throws {
    let (useCase, payments, _, _) = try pipeline(stock: ["SKU-1": 10, "SKU-2": 10])
    let first = useCase.execute(
      lines: [try line("SKU-1", quantity: 1, minorUnits: 100)],
      idempotencyKey: "idem-8"
    )
    let second = useCase.execute(
      lines: [try line("SKU-2", quantity: 1, minorUnits: 100)],
      idempotencyKey: "idem-8"
    )
    guard case .success(_) = first else {
      Issue.record("expected first success, got \(first)")
      return
    }
    guard case .failure(.invalidOrder(let error)) = second else {
      Issue.record("expected invalid order on reuse, got \(second)")
      return
    }
    #expect(error.reason.contains("idempotency"))
    #expect(payments.chargedOrders.count == 1)
  }

  @Test func clockPortIsInjectable() {
    let clock = FixedClock(1_700_000_000_000)
    #expect(clock.now() == 1_700_000_000_000)
  }
}

File: csharp/tests/Warehouse.IntegrationTests/PlaceOrderTests.cs
using Warehouse.Adapters;
using Warehouse.Application;
using Warehouse.Domain;

namespace Warehouse.IntegrationTests;

public sealed class PlaceOrderTests
{
    private sealed record Pipeline(
        PlaceOrderUseCase UseCase,
        FakePaymentProcessor Payments,
        InMemoryOrderRepository Repository,
        InMemoryInventoryGateway Inventory);

    private static Pipeline Create(Dictionary<string, int> stock, bool decline = false, string orderId = "ord-1")
    {
        var initial = stock.ToDictionary(pair => new Sku(pair.Key), pair => pair.Value);
        var inventory = new InMemoryInventoryGateway(initial);
        var payments = new FakePaymentProcessor(decline);
        var repository = new InMemoryOrderRepository();
        var useCase = new PlaceOrderUseCase(
            inventory,
            payments,
            repository,
            new FixedOrderIdGenerator(new OrderId(orderId)));
        return new Pipeline(useCase, payments, repository, inventory);
    }

    private static List<OrderLine> Lines(params (string Code, int Qty, long MinorUnits)[] specs)
    {
        return specs
            .Select(spec => new OrderLine(
                new Sku(spec.Code),
                new Quantity(spec.Qty),
                new Money(spec.MinorUnits, "USD")))
            .ToList();
    }

    [Fact]
    public void HappyPathPersistsPaid()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10 });
        PlaceOrderResult result = pipeline.UseCase.Execute(Lines(("SKU-1", 2, 500)), "idem-1");

        var success = Assert.IsType<PlaceOrderResult.Success>(result);
        Assert.Equal(OrderStatus.Paid, success.Order.Status);
        Assert.Equal("ord-1", success.Order.Id.Value);
        Assert.Equal(new Money(1000, "USD"), success.Order.Total());
        Assert.Single(pipeline.Payments.ChargedOrders);
        Order? stored = pipeline.Repository.Get(success.Order.Id);
        Assert.NotNull(stored);
        Assert.Equal(OrderStatus.Paid, stored.Status);
        Assert.NotSame(stored, success.Order);
        Assert.Equal(8, pipeline.Inventory.SnapshotStock()[new Sku("SKU-1")]);
    }

    [Fact]
    public void InsufficientStockFailsWithoutChargeOrPersist()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 1 });
        PlaceOrderResult result = pipeline.UseCase.Execute(Lines(("SKU-1", 5, 500)), "idem-2");

        var failure = Assert.IsType<PlaceOrderResult.Failure>(result);
        var stock = Assert.IsType<InsufficientStock>(failure.Error);
        Assert.Equal(1, stock.Available);
        Assert.Empty(pipeline.Payments.ChargedOrders);
        Assert.Empty(pipeline.Repository.Saved);
        Assert.Equal(1, pipeline.Inventory.SnapshotStock()[new Sku("SKU-1")]);
    }

    [Fact]
    public void PaymentDeclineReleasesReservation()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10 }, decline: true);
        PlaceOrderResult result = pipeline.UseCase.Execute(Lines(("SKU-1", 1, 500)), "idem-3");

        var failure = Assert.IsType<PlaceOrderResult.Failure>(result);
        Assert.IsType<PaymentDeclined>(failure.Error);
        Assert.Single(pipeline.Payments.ChargedOrders);
        Assert.Empty(pipeline.Repository.Saved);
        Assert.Equal(10, pipeline.Inventory.SnapshotStock()[new Sku("SKU-1")]);
    }

    [Fact]
    public void SaveFailureRefundsAndReleases()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10 });
        pipeline.Repository.FailSave = true;
        PlaceOrderResult result = pipeline.UseCase.Execute(Lines(("SKU-1", 1, 500)), "idem-4");

        var failure = Assert.IsType<PlaceOrderResult.Failure>(result);
        Assert.IsType<PersistenceConflict>(failure.Error);
        Assert.Single(pipeline.Payments.Refunded);
        Assert.Equal(10, pipeline.Inventory.SnapshotStock()[new Sku("SKU-1")]);
    }

    [Fact]
    public void CompensationFailureAfterSaveFailure()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10 });
        pipeline.Repository.FailSave = true;
        pipeline.Payments.FailRefund = true;
        PlaceOrderResult result = pipeline.UseCase.Execute(Lines(("SKU-1", 1, 500)), "idem-5");

        var failure = Assert.IsType<PlaceOrderResult.Failure>(result);
        Assert.IsType<CompensationFailure>(failure.Error);
    }

    [Fact]
    public void InvalidLinesFailValidation()
    {
        Pipeline pipeline = Create([]);
        PlaceOrderResult result = pipeline.UseCase.Execute([], "idem-6");

        var failure = Assert.IsType<PlaceOrderResult.Failure>(result);
        Assert.IsType<InvalidOrderException>(failure.Error);
    }

    [Fact]
    public void GetReturnsNullForUnknownId()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10 });
        Assert.Null(pipeline.Repository.Get(new OrderId("missing-order")));
    }

    [Fact]
    public void IdempotentReplayDoesNotDoubleCharge()
    {
        var inventory = new InMemoryInventoryGateway(new Dictionary<Sku, int> { [new Sku("SKU-1")] = 10 });
        var payments = new FakePaymentProcessor();
        var repository = new InMemoryOrderRepository();
        var useCase = new PlaceOrderUseCase(inventory, payments, repository, new SequenceOrderIdGenerator());
        PlaceOrderResult first = useCase.Execute(Lines(("SKU-1", 2, 300)), "idem-7");
        Assert.IsType<PlaceOrderResult.Success>(first);
        PlaceOrderResult second = useCase.Execute(Lines(("SKU-1", 2, 300)), "idem-7");
        Assert.IsType<PlaceOrderResult.Success>(second);
        Assert.Single(payments.ChargedOrders);
    }

    [Fact]
    public void ReusedKeyWithDifferentPayloadIsInvalid()
    {
        Pipeline pipeline = Create(new Dictionary<string, int> { ["SKU-1"] = 10, ["SKU-2"] = 10 });
        Assert.IsType<PlaceOrderResult.Success>(pipeline.UseCase.Execute(Lines(("SKU-1", 1, 100)), "idem-8"));
        PlaceOrderResult second = pipeline.UseCase.Execute(Lines(("SKU-2", 1, 100)), "idem-8");
        var failure = Assert.IsType<PlaceOrderResult.Failure>(second);
        Assert.IsType<InvalidOrderException>(failure.Error);
        Assert.Contains("idempotency key reused", failure.Error.Message, StringComparison.Ordinal);
    }
}
